packages/docusaurus-utils/src/index.ts

```diff
@@ -252,43 +252,71 @@ export function createExcerpt(fileString: string): string | undefined {
 }
 
 type ParsedMarkdown = {
-  frontMatter: {
-    // Returned by gray-matter
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    [key: string]: any;
-  };
+  // Returned by gray-matter
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  frontMatter: Record<string, any>;
   content: string;
   excerpt: string | undefined;
+  hasFrontMatter: boolean;
 };
-export function parseMarkdownString(markdownString: string): ParsedMarkdown {
-  const options: Record<string, unknown> = {
-    excerpt: (file: matter.GrayMatterFile<string>): void => {
-      // Hacky way of stripping out import statements from the excerpt
-      // TODO: Find a better way to do so, possibly by compiling the Markdown content,
-      // stripping out HTML tags and obtaining the first line.
-      file.excerpt = createExcerpt(file.content);
-    },
-  };
 
+export function readFrontMatter(
+  markdownString: string,
+  source?: string,
+  options: Record<string, unknown> = {},
+  removeTitleHeading?: boolean,
+): ParsedMarkdown {
   try {
-    const {data: frontMatter, content, excerpt} = matter(
-      markdownString,
-      options,
-    );
-    return {frontMatter, content, excerpt};
+    const result = matter(markdownString, options);
+    result.data = result.data || {};
+
+    const hasFrontMatter = Object.keys(result.data).length > 0;
+
+    const heading = /^# (.*)[\n\r]/gi.exec(result.content.trim());
+    if (heading) {
+      if (removeTitleHeading) {
+        result.content = result.content.replace(heading[0], '');
+      }
+      if (result.data.title) {
+        console.warn(`Duplicate title detected in ${source || 'this file'}`);
+      } else {
+        result.data.title = heading[1].trim();
+      }
```

Comment:
i'm not sure if we should always remove heading for docs and blog or do this only when its not duplicated
```ts
      if (result.data.title) {
        console.warn(`Duplicate title detected in ${source || 'this file'}`);
      } else {
        result.data.title = heading[1].trim();
        if (removeTitleHeading) {
          result.content = result.content.replace(heading[0], '');
        }
      }
```